CODING_GUIDELINES.md

```diff
@@ -1,13 +1,140 @@
+* `mypackage_test.go` contains tests for your package.
+  If you like, you can use the `package mypackage_test` in that file, however, this is up to you.
+* `error.go` contains sentinel errors.
+  For possible content, see section _Sentinel errors_ below.
+* Any other file in your package that is necessary.
+  If you use a lot of structs with methods, try to place each struct in a separate file, that is named like the struct, but in lower-snake-case (`hello_world`).
+
+## Code
+We want to keep our code readable (that's one of the main points of this project) and clean.
+Many of the things we expect from code, is ensured by our internal analyzers (`internal/tool/analysis`).
+However, not everything may be covered by those.
+This is why we rely on you to follow the below rules for our code.
+
+### Ordering
+We want to keep the following structure in any non-test `.go`-file.
+
+1. package declaration (no package comment, use a file `doc.go` for that)
+2. import declarations
+3. const declarations
+4. var declarations
+5. interface declarations
+6. struct declarations (followed by constructor functions)
+7. exported methods
+8. unexported methods
+9. exported functions
+10. unexported functions
+
+If there are multiple independent structs in one file, you should probably split the file into two or more.
+
+### Constructor functions
+Use the `&struct` syntax for initialization whenever possible.
+If you need to do previous initialization, do this before inline-initializing the struct in the return statement.
+The example below describes a correct example of how to write a constructor.
+
+```go
+type Foo struct {
+    id   [16]byte
+    x    int
+    y    int
+    desc string
+}
+func NewFoo(desc string) *Foo {
+    id := createID()
+    return &Foo{
+        id:   id,
+        x:    0,
+        y:    0,
+        desc: desc
+    }
+}
+
+```
+
+### `context.Context`
+If not in a performance critical layer, use `context.Context` and respect it.
+Especially when writing network APIs, `context.Context` is a must.
+When you use it, there are a few rules.
+
+* a `context.Context` must always be the first argument in a method or function
+* there must be only one `context.Context` argument
+* the `context.Context` argument has to be named `ctx`
+* the context has to be respected (implementation must abort as soon as possible when `<-ctx.Done()` returns)
+
+### Sentinel errors
+Sentinel errors are exported error values that you use for comparison.
+```go
+res, err := somepkg.Operation()
+if err == somepkg.ErrTimeout {
+    // timeout, maybe try again?
+} else if err != nil {
+    // some other error, handle accordingly
+}
+```
+If you make use of sentinel errors in your package, **sentinel errors must be constant**.
+Go itself doesn't follow this rule.
+As an example, `io.EOF` is not a constant error.
+Run a larger application after calling `io.EOF = nil` and see what happens.
```

Comment:
Was this supposed to be `io.EOF == nil` ?

Reply:
No. `io.EOF` is not constant, and you can assign `nil` to it.